sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/SdkRequestContext.java

```diff
@@ -0,0 +1,157 @@
+import io.clientcore.core.utils.ProgressReporter;
+
+import java.util.function.Consumer;
+
+/**
+ * This class contains the request metadata that flows along with the {@link HttpRequest} as it goes through HTTP pipeline.
+ * It includes {@link RequestOptions} provided by the user along with metadata provided by client libraries and core components.
+ * <p>
+ * This class is intended to be used by client library code and HTTP pipeline components. Application developers should
+ * configure request metadata using {@link RequestOptions}.
+ * <p>
+ * This class does not allow any modifications to the request metadata provided by the user.
+ */
+public final class SdkRequestContext extends RequestOptions {
+    // TODO (limolkova) do we need to make it public and probably move to another class?
+    private static final String PROGRESS_REPORTER_KEY = "progressReporter";
+    static final SdkRequestContext NONE = new SdkRequestContext(null, null, null);
+    private static final ClientLogger LOGGER = new ClientLogger(SdkRequestContext.class);
+    private final InstrumentationContext childInstrumentationContext;
+    private final ProgressReporter childProgressReporter;
+
+    private SdkRequestContext(RequestOptions options, InstrumentationContext instrumentationContext,
+        ProgressReporter progressReporter) {
+        super(options);
+
+        this.childInstrumentationContext = instrumentationContext;
+        this.childProgressReporter = progressReporter;
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options and instrumentation context.
+     *
+     * @param options The {@link RequestOptions} to be used.
+     * @param instrumentationContext The {@link InstrumentationContext} to be used.
+     * @return A new instance of {@link SdkRequestContext} or, when optimization is possible, the existing instance of {@link SdkRequestContext}
+     * provided in the {@code options} parameter.
+     */
+    public static SdkRequestContext create(RequestOptions options, InstrumentationContext instrumentationContext) {
+        if (options instanceof SdkRequestContext
+            && (instrumentationContext == null || !instrumentationContext.isValid())) {
+            return (SdkRequestContext) options;
+        }
+
+        return new SdkRequestContext(options, instrumentationContext, null);
+    }
+
+    /**
+     * Creates a new instance of {@link SdkRequestContext} with the given options and progress reporter.
+     *
+     * @param options The {@link RequestOptions} to be used.
+     * @param progressReporter The {@link ProgressReporter} to be used.
+     * @return A new instance of {@link SdkRequestContext} or, when optimization is possible, the existing instance of {@link SdkRequestContext}
+     * provided in the {@code options} parameter.
+     */
+    public static SdkRequestContext create(RequestOptions options, ProgressReporter progressReporter) {
```

Comment:
Is there a case where we need to set all three of them - RequestOptions, ProgressReporter and InstrumentationContext? We don't have a factory method to do that.

Reply:
changes to copy-on-write, based  on current storage design, we'd have to set progressReporter independently from InstrumentationContext. 

I imagine we might want to play with it in the storage v2 SDK a bit before making a final call on this API. I'd prefer to merge and polish.